src/jit/decomposelongs.cpp

```diff
@@ -964,20 +975,252 @@ GenTree* DecomposeLongs::DecomposeShift(LIR::Use& use)
             unreached();
     }
 
-    GenTreeArgList* argList = m_compiler->gtNewArgList(loOp1, hiOp1, shiftWidthOp);
+    // If we are shifting by a constant int, we do not want to save to a helper, instead, we decompose.
+    if (shiftOper == GT_CNS_INT)
+    {
+        GenTree* loCopy  = m_compiler->gtNewLclvNode(loOp1LclNum, TYP_INT);
 
-    GenTree* call = m_compiler->gtNewHelperCallNode(helper, TYP_LONG, 0, argList);
+        GenTree* loResult;
+        GenTree* hiResult;
 
-    GenTreeCall*    callNode    = call->AsCall();
-    ReturnTypeDesc* retTypeDesc = callNode->GetReturnTypeDesc();
-    retTypeDesc->InitializeLongReturnType(m_compiler);
+        switch (oper)
+        {
+            case GT_LSH:
+            {
+                if (count == 0)
+                {
+                    // Do nothing.
+                    loResult = loOp1;
+                    hiResult = hiOp1;
+                    Range().InsertAfter(tree, loResult, hiResult);
+                }
+                else if (count < 32)
+                {
+                    // Hi is a GT_LSH_HI, lo is a GT_LSH. Will produce:
+                    // reg1 = lo
+                    // shl lo, shift
+                    // shld hi, reg1, shift
+
```

Comment:
I think we need to fix FinalizeDecomposition() to take an insertAfter argument, and not assert anything about the relative ordering of loResult and hiResult. Then I think you could simply insert the high nodes before the low nodes in linear order. You do need an additional lclvar node, so you do need loCopy (it's not actually a copy of the value, but an additional lclvar reference). You need to call `m_compiler->lvaIncRefCnts(loCopy);` to account for the additional lclvar reference in the LSH_HI tree.


---
In reply to: [78656652](https://github.com/dotnet/coreclr/pull/7169#discussion_r78656652) [](ancestors = 78656652)